### DISCUSSION TITLE
Why does Saleor generate a new address entry for every order?

### DISCUSSION THREAD

**Discussion by peteror** [saleor/saleor#12380] [Q&A]
I have noticed that Saleor creates a new entry in the account_address table for every order with shipping, even though my customers have saved addresses. 

At first I thought my implementation (using a mobile app as the frontend) was wrong, using checkoutShippingAddressUpdate to add the address to the checkout. It turns out checkoutCreate only takes AddressInput (with all the address fields) but not the ID of a saved address.  

I already have about 2.6 million lines in my account_address table - is there a way around it / is it something on the roadmap to fix?

**Comment by mociepka** [comment]
Hi, the order billing address can't be changed even if your customer changes it in the future.

Theoretically, if you reuse the same record, all historical order billing data will change when the user changes the default billing address. Your finance people won't be happy ;)

Shipping is not crucial, but it could confuse your staff if someone changes the default shipping address but 3rd party tracking system gets the old one.

That being said, you need to store (even duplicate) order addresses. Is the `account_address` table a good place? I would rather store it on a separate table or as a part of the order. I can imagine that your address picker is very slow right now.

**Comment by peteror** [comment]
That makes sense, although this way, I end up duplicating (many times over) the addresses at our external shipper's system, as I create a new address with every order there too. 

Interestingly, my address picker is not too slow, as the `{ me { addresses } }` array returns only the addresses I specifically saved: these seem to be linked to the user account in the `account_user_addresses` table. But when I re-use any of these, the checkout object gives me the order ID it saves for the order, which is a copy of one of the originals I saved... 

It may be a good idea to return the saved address ID as well when re-using one, (mostly important for external integrations I think) while linking the copy to the order so that it cannot be changed.

**Comment by patrys** [comment]
Reading the above discussion makes me wonder if we could improve developer experiences by making addresses copy-on-write instead? This would require a change to how address events work but would prevent having duplicate address entries.
